### ISSUE TITLE
ebs_block_device parameter 'encrypted' evaluated as 'true' even when not set

### ISSUE THREAD

**Issue by gilmatic** [hashicorp/terraform#5360]
When attempting to include an ebs_block_device with an encrypted snapshot in an aws_launch_configuration, the following error is received no matter how the ebs_block_device is defined:

_Parameter encrypted is invalid. You cannot specify the encrypted flag if specifying a snapshot id in a block device mapping. Launching EC2 instance failed._

The ebs_block_device gave the same error whether configured any of the following three  ways:

```
 ebs_block_device {
    encrypted = 0
    device_name = "/dev/xvdg"
    volume_size = "50"
    snapshot_id = "snap-1234567"
    delete_on_termination = "true"
  }
```

or

```
 ebs_block_device {
    encrypted = 1
    device_name = "/dev/xvdg"
    volume_size = "50"
    snapshot_id = "snap-1234567"
    delete_on_termination = "true"
  }
```

or 

```
 ebs_block_device {
    device_name = "/dev/xvdg"
    volume_size = "50"
    snapshot_id = "snap-1234567"
    delete_on_termination = "true"
  }
```

When set to '0' tfstate reads the following way:

  ebs_block_device.#:                                "" => "1"
  ebs_block_device.2403880458.delete_on_termination: "" => "1"
  ebs_block_device.2403880458.device_name:           "" => "/dev/xvdg"
  **ebs_block_device.2403880458.encrypted:             "" => "0"**
  ebs_block_device.2403880458.iops:                  "" => "<computed>"
  ebs_block_device.2403880458.snapshot_id:           "" => "snap-1234567"
  ebs_block_device.2403880458.volume_size:           "" => "50"
  ebs_block_device.2403880458.volume_type:           "" => "<computed>"
  ebs_optimized:                                     "" => "<computed>"

When set to '1' or *undefined tfstate shows the encrypted value as '1'.

The AWS API appears to not accept any encrypted value if a snapshot used for the ebs_block_device.

If a snapshot is specified in an ebs_block_device, terraform should ignore the encrypted value and not expose it to the API, which would hopefully allow the autoscaling group to spin up instances successfully.

**Comment by buildmaster** [comment]
just ran into this too, seems related to #1862 in that the api should be passing nil in the launch configuration? Occurs in version 0.6.12 and 0.6.11.

**Comment by gilmatic** [comment]
Last known good in our enterprise was 0.6.8 and the issue does not exist in this version...somewhere between 9-12 is where the error arose.

**Comment by anthcor** [comment]
:+1:  working on this with @gilmatic too

**Comment by xsmaster** [comment]
I can provide a bit more information on this issue. Version 0.6.9 works correctly with the aws_launch_configuration seen above from @gilmatic I use the same thing in my code. 

The ASG Fails to bring an instance in service due to the encrypted pram. 

Terraform versions 0.6.10+ does the below.
```
  aws_launch_configuration.launch_config: Creating...
    associate_public_ip_address:                       "" => "0"
    ebs_block_device.#:                                "" => "1"
    ebs_block_device.2403880458.delete_on_termination: "" => "1"
    ebs_block_device.2403880458.device_name:           "" => "/dev/xvdg"
    ebs_block_device.2403880458.encrypted:             "" => "<computed>"
    ebs_block_device.2403880458.iops:                  "" => "<computed>"
    ebs_block_device.2403880458.snapshot_id:           "" => "snap-1234567"
    ebs_block_device.2403880458.volume_size:           "" => "50"
    ebs_block_device.2403880458.volume_type:           "" => "gp2"
    ebs_optimized:                                     "" => "<computed>"
```
Terraform version 0.6.12 State file:
```
"attributes": {
                            "associate_public_ip_address": "false",
                            "ebs_block_device.#": "1",
                            "ebs_block_device.2403880458.delete_on_termination": "true",
                            "ebs_block_device.2403880458.device_name": "/dev/xvdg",
                            "ebs_block_device.2403880458.encrypted": "false"
                            "ebs_block_device.2403880458.iops": "0",
                            "ebs_block_device.2403880458.snapshot_id": "snap-1234567",
                            "ebs_block_device.2403880458.volume_size": "50",
                            "ebs_block_device.2403880458.volume_type": "gp2",
                            "ebs_optimized": "false",
                            ...
``` 
Terraform version 0.6.9 I see the below and works.
```
  aws_launch_configuration.launch_config: Creating...
    associate_public_ip_address:                       "" => "0"
    ebs_block_device.#:                                "" => "1"
    ebs_block_device.2403880458.delete_on_termination: "" => "1"
    ebs_block_device.2403880458.device_name:           "" => "/dev/xvdg"
    ebs_block_device.2403880458.iops:                  "" => "<computed>"
    ebs_block_device.2403880458.snapshot_id:           "" => "snap-1234567"
    ebs_block_device.2403880458.volume_size:           "" => "50"
    ebs_block_device.2403880458.volume_type:           "" => "gp2"
    ebs_optimized:                                     "" => "<computed>"
```

Trace Log:
```
# LOG
016/03/08 06:48:34 [DEBUG] apply: aws_launch_configuration.launch_config: executing Apply
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws: 2016/03/08 06:48:34 [DEBUG] autoscaling create launch configuration: {
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:   BlockDeviceMappings: [{
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:       DeviceName: "/dev/xvdg",
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:       Ebs: {
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:         DeleteOnTermination: true,
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:         Encrypted: false,
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:         SnapshotId: "snap-12345678",
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:         VolumeSize: 50,
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:         VolumeType: "gp2"
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:       }
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:     }],
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:   EbsOptimized: false,
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:   ImageId: "ami-12345678",
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:   InstanceMonitoring: {
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:     Enabled: true
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:   },
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:   InstanceType: "t2.large",
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:   LaunchConfigurationName: "Test-lc",
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws:   SecurityGroups: ["sg-12345678"],
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws: }
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws: 2016/03/08 06:48:34 [DEBUG] Waiting for state to become: [success]
2016/03/08 06:48:34 [DEBUG] terraform-provider-aws: 2016/03/08 06:48:34 [TRACE] Waiting 500ms before next try
2016/03/08 06:48:36 [DEBUG] terraform-provider-aws: 2016/03/08 06:48:36 [INFO] launch configuration ID: Test-lc
2016/03/08 06:48:36 [DEBUG] terraform-provider-aws: 2016/03/08 06:48:36 [DEBUG] Waiting for state to become: [success]
2016/03/08 06:48:36 [DEBUG] terraform-provider-aws: 2016/03/08 06:48:36 [TRACE] Waiting 500ms before next try
2016/03/08 06:48:36 [DEBUG] terraform-provider-aws: 2016/03/08 06:48:36 [DEBUG] launch configuration describe configuration: {
2016/03/08 06:48:36 [DEBUG] terraform-provider-aws:   LaunchConfigurationNames: ["Test-lc"]
2016/03/08 06:48:36 [DEBUG] terraform-provider-aws: }
```

**Comment by xsmaster** [comment]
Found a PR 4481 https://github.com/hashicorp/terraform/pull/4481 that caused this issue for the Launch config. I could be wrong since I don't know go yet but I don't see a check for snapshot_id. If there is a snapshot_id then the ebs_block_device.2403880458.encrypted should not show up.

**Comment by bryanvaz** [comment]
The issue seems also be inconsistent on the AWS side. If you specify a spot price in exact same code as above, the problem will go away, and the ASG can spin up an instance fine, even if it has and 'encrypted false' flag.

**Comment by silverjava** [comment]
is there anyone who knows how to get around of it?

**Comment by xsmaster** [comment]
Going back to an earlier version is what I had to do. Not the best plan. :(

**Comment by MiguelMoll** [comment]
Checking in to see if there's been any progress or perhaps another work around (besides downgrading) for this problem.

**Comment by silverjava** [comment]
I was doing downgrade to work around of it :(

**Comment by xsmaster** [comment]
It's what I did.

**Comment by jen20** [comment]
Fixed in #6452. Thanks for the detailed report!
